tsne/bh_sne_src/tsne.cpp

```diff
@@ -56,6 +57,16 @@ using std::move;
 using std::sqrt;
 using std::vector;
 
+struct TSNEState {
+  std::vector<double> P;
+  std::vector<unsigned int> row_P;
+  std::vector<unsigned int> col_P;
+  std::vector<double> val_P;
+  std::vector<double> dY;
+  std::vector<double> uY;
+  std::vector<double> gains;
+};
```

Comment:
Move this inside the anonymous namespace, or move it into the header.  Also, you have `using std::vector` so
```suggestion
struct TSNEState {
  vector<double> P;
  vector<unsigned int> row_P;
  vector<unsigned int> col_P;
  vector<double> val_P;
  vector<double> dY;
  vector<double> uY;
  vector<double> gains;
};
```